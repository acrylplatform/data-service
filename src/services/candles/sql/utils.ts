import { findLast } from 'ramda';
import { Result, Ok as ok, Error as error } from 'folktale/result';
import { Interval } from '../../../types';
import { div } from '../../../utils/interval';
import { ValidationError } from '../../../errorHandling';

export const highestDividerLessThan = (
  inter: Interval,
  dividers: Interval[]
): Result<ValidationError, Interval> => {
  const i = findLast((i: Interval) => div(inter, i) >= 1, dividers);
  return i ? ok(i) : error(new ValidationError('Divider not found'));
};

/**
 * Composes sum using a polynom with additives from given units array
 * Uses a greedy algorithm
 * @param units is sorted desc
 * @param sum
 */
export const numberToUnitsPolynom = (
  units: number[],
  sum: number
): number[][] => {
  const [, result] = units.reduce<[number, number[][]]>(
    ([remaining, result], unit) => {
      const k = Math.floor(remaining / unit);
      if (k > 0) {
        return [remaining - k * unit, [...result, [unit, k]]];
      } else {
        return [remaining, result];
      }
    },
    [sum, []]
  );

  return result;
};
