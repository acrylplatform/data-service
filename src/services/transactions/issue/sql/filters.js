const { where } = require('../../../../utils/db/knex');

const commonFilters = require('../../_common/sql/filters');
const commonFiltersOrder = require('../../_common/sql/filtersOrder');

const script = where('script');
const assetId = where('asset_id');

module.exports = {
  filters: {
    ...commonFilters,
    script,
    assetId,
  },
  filtersOrder: [...commonFiltersOrder, 'script', 'assetId'],
};
