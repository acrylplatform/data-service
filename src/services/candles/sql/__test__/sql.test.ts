import { QueryBuilder } from 'knex';
import { unsafeIntervalsFromStrings } from '../../../../utils/interval';
import { periodsToQueries, sql } from '../sql';

describe('sql queries from period', () => {
  it('should return empty array', () => {
    expect(
      periodsToQueries({
        amountAsset: '111',
        priceAsset: '222',
        timeStart: new Date('2019-03-25T00:00:00.000Z'),
        periods: [],
        matcher: '123',
      }).length
    ).toBe(0);
  });

  it('should return array with 3 queries (30m, 5m, 1m)', () => {
    const periodQueries = periodsToQueries({
      amountAsset: '111',
      priceAsset: '222',
      timeStart: new Date('2019-03-25T00:00:00.000Z'),
      periods: unsafeIntervalsFromStrings(['30m', '5m', '1m']),
      matcher: '123',
    });
    expect(periodQueries.length).toBe(3);
    expect(
      periodQueries.reduce(
        (sql: string, q: QueryBuilder) => `${sql} UNION ${q.toString()}`,
        ''
      )
    ).toMatchSnapshot();
  });

  it('should return array with 5 queries', () => {
    const periodQueries = periodsToQueries({
      amountAsset: '111',
      priceAsset: '222',
      timeStart: new Date('2019-03-25T00:00:00.000Z'),
      periods: unsafeIntervalsFromStrings(['30m', '15m', '1m', '1m', '1m']),
      matcher: '123',
    });
    expect(periodQueries.length).toBe(5);
    expect(
      periodQueries.reduce(
        (sql: string, q: QueryBuilder) => `${sql} UNION ${q.toString()}`,
        ''
      )
    ).toMatchSnapshot();
  });
});

describe('sql query from candles', () => {
  it('should search candles for 1h', () => {
    expect(
      sql({
        amountAsset: '111',
        priceAsset: '222',
        timeStart: new Date('2017-04-03T00:00:00.000Z'),
        timeEnd: new Date('2017-04-03T23:59:59.999Z'),
        interval: '1h',
        matcher: '123',
      })
    ).toMatchSnapshot();
  });

  it('should search candles for 1d', () => {
    expect(
      sql({
        amountAsset: '111',
        priceAsset: '222',
        timeStart: new Date('2017-04-03T00:00:00.000Z'),
        timeEnd: new Date('2017-04-03T23:59:59.999Z'),
        interval: '1d',
        matcher: '123',
      })
    ).toMatchSnapshot();
  });

  it('should search candles for 1m', () => {
    expect(
      sql({
        amountAsset: '111',
        priceAsset: '222',
        timeStart: new Date('2017-04-03T00:00:00.000Z'),
        timeEnd: new Date('2017-04-03T23:59:59.999Z'),
        interval: '1m',
        matcher: '123',
      })
    ).toMatchSnapshot();
  });
});
