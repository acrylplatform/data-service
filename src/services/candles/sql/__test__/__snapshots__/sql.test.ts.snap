// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql queries from period should return array with 3 queries (30m, 5m, 1m) 1`] = `" UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:00:00.000' and \\"time_start\\" <= '2019-03-25 03:30:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 1800 limit 1 UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:30:00.000' and \\"time_start\\" <= '2019-03-25 03:35:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 300 limit 1 UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:35:00.000' and \\"time_start\\" <= '2019-03-25 03:36:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1"`;

exports[`sql queries from period should return array with 5 queries 1`] = `" UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:00:00.000' and \\"time_start\\" <= '2019-03-25 03:30:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 1800 limit 1 UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:30:00.000' and \\"time_start\\" <= '2019-03-25 03:45:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 900 limit 1 UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:45:00.000' and \\"time_start\\" <= '2019-03-25 03:46:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1 UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:46:00.000' and \\"time_start\\" <= '2019-03-25 03:47:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1 UNION select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2019-03-25 03:47:00.000' and \\"time_start\\" <= '2019-03-25 03:48:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1"`;

exports[`sql query from candles should search candles for 1d 1`] = `"select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\", \\"a_dec\\".\\"decimals\\" as \\"a_dec\\", \\"p_dec\\".\\"decimals\\" as \\"p_dec\\" from (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-03 03:00:00.000' and \\"time_start\\" <= '2017-04-04 02:59:59.999' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 86400 union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-03 03:00:00.000' and \\"time_start\\" <= '2017-04-03 15:00:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 43200 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-03 15:00:00.000' and \\"time_start\\" <= '2017-04-03 21:00:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 21600 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-03 21:00:00.000' and \\"time_start\\" <= '2017-04-04 00:00:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 10800 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 00:00:00.000' and \\"time_start\\" <= '2017-04-04 01:00:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 3600 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 01:00:00.000' and \\"time_start\\" <= '2017-04-04 02:00:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 3600 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:00:00.000' and \\"time_start\\" <= '2017-04-04 02:30:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 1800 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:30:00.000' and \\"time_start\\" <= '2017-04-04 02:45:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 900 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:45:00.000' and \\"time_start\\" <= '2017-04-04 02:50:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 300 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:50:00.000' and \\"time_start\\" <= '2017-04-04 02:55:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 300 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:55:00.000' and \\"time_start\\" <= '2017-04-04 02:56:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:56:00.000' and \\"time_start\\" <= '2017-04-04 02:57:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:57:00.000' and \\"time_start\\" <= '2017-04-04 02:58:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:58:00.000' and \\"time_start\\" <= '2017-04-04 02:59:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1)) as \\"c\\" inner join \\"asset_decimals\\" as \\"a_dec\\" on \\"c\\".\\"amount_asset_id\\" = \\"a_dec\\".\\"asset_id\\" inner join \\"asset_decimals\\" as \\"p_dec\\" on \\"c\\".\\"price_asset_id\\" = \\"p_dec\\".\\"asset_id\\" order by \\"c\\".\\"time_start\\" asc"`;

exports[`sql query from candles should search candles for 1h 1`] = `"select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\", \\"a_dec\\".\\"decimals\\" as \\"a_dec\\", \\"p_dec\\".\\"decimals\\" as \\"p_dec\\" from (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-03 03:00:00.000' and \\"time_start\\" <= '2017-04-04 02:59:59.999' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 3600 union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:00:00.000' and \\"time_start\\" <= '2017-04-04 02:30:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 1800 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:30:00.000' and \\"time_start\\" <= '2017-04-04 02:45:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 900 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:45:00.000' and \\"time_start\\" <= '2017-04-04 02:50:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 300 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:50:00.000' and \\"time_start\\" <= '2017-04-04 02:55:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 300 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:55:00.000' and \\"time_start\\" <= '2017-04-04 02:56:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:56:00.000' and \\"time_start\\" <= '2017-04-04 02:57:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:57:00.000' and \\"time_start\\" <= '2017-04-04 02:58:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1) union (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-04 02:58:00.000' and \\"time_start\\" <= '2017-04-04 02:59:00.000' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60 limit 1)) as \\"c\\" inner join \\"asset_decimals\\" as \\"a_dec\\" on \\"c\\".\\"amount_asset_id\\" = \\"a_dec\\".\\"asset_id\\" inner join \\"asset_decimals\\" as \\"p_dec\\" on \\"c\\".\\"price_asset_id\\" = \\"p_dec\\".\\"asset_id\\" order by \\"c\\".\\"time_start\\" asc"`;

exports[`sql query from candles should search candles for 1m 1`] = `"select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\", \\"a_dec\\".\\"decimals\\" as \\"a_dec\\", \\"p_dec\\".\\"decimals\\" as \\"p_dec\\" from (select \\"time_start\\", \\"amount_asset_id\\", \\"price_asset_id\\", \\"low\\", \\"high\\", \\"volume\\", \\"quote_volume\\", \\"max_height\\", \\"txs_count\\", \\"weighted_average_price\\", \\"open\\", \\"close\\", \\"interval_in_secs\\", \\"matcher\\" from \\"candles\\" where \\"amount_asset_id\\" = '111' and \\"price_asset_id\\" = '222' and \\"time_start\\" >= '2017-04-03 03:00:00.000' and \\"time_start\\" <= '2017-04-04 02:59:59.999' and \\"matcher\\" = '123' and \\"interval_in_secs\\" = 60) as \\"c\\" inner join \\"asset_decimals\\" as \\"a_dec\\" on \\"c\\".\\"amount_asset_id\\" = \\"a_dec\\".\\"asset_id\\" inner join \\"asset_decimals\\" as \\"p_dec\\" on \\"c\\".\\"price_asset_id\\" = \\"p_dec\\".\\"asset_id\\" order by \\"c\\".\\"time_start\\" asc"`;
